lib/chef-dk/policyfile/git_lock_fetcher.rb

```diff
@@ -0,0 +1,252 @@
+      #
+      # @param name [String] The name of the policyfile
+      # @param source_options [Hash] A hash with a :path key pointing at the location
+      #                              of the lock
+      def initialize(name, source_options, storage_config)
+        @name           = name
+        @storage_config = storage_config
+        @source_options = source_options
+
+        @uri      = source_options[:git]
+        @branch   = source_options[:branch]
+        @tag      = source_options[:tag]
+        @ref      = source_options[:ref]
+        @revision = source_options[:revision]
+        @path     = source_options[:path] || source_options[:rel]
+
+        # The revision to parse
+        @rev_parse = source_options[:ref] || source_options[:branch] || source_options[:tag] || "master"
+      end
+
+      # @return [True] if there were no errors with the provided source_options
+      # @return [False] if there were errors with the provided source_options
+      def valid?
+        errors.empty?
+      end
+
+      # Check the options provided when craeting this class for errors
+      #
+      # @return [Array<String>] A list of errors found
+      def errors
+        error_messages = []
+        [:git].each do |key|
+          error_messages << "include_policy for #{name} is missing key #{key}" unless source_options[key]
+        end
+
+        error_messages
+      end
+
+      # @return [Hash] The source_options that describe how to fetch this exact lock again
+      def source_options_for_lock
+        source_options.merge({
+                               revision:  revision,
+                             })
+      end
+
+      # Applies source options from a lock file. This is used to make sure that the same
+      # policyfile lock is loaded that was locked
+      #
+      # @param options_from_lock [Hash] The source options loaded from a policyfile lock
+      def apply_locked_source_options(options_from_lock)
+        # There are no options the lock could provide
+
+        options = options_from_lock.inject({}) do |acc, (key, value)|
+          acc[key.to_sym] = value
+          acc
+        end
+        source_options.merge!(options)
+        raise ChefDK::InvalidLockfile, "Invalid source_options provided from lock data: #{options_from_lock_file.inspect}" if !valid?
+      end
+
+      # @return [String] of the policyfile lock data
+      def lock_data
+        @lock_data ||= fetch_lock_data.tap do |data|
+          data["cookbook_locks"].each do |cookbook_name, cookbook_lock|
+            cookbook_path = cookbook_lock["source_options"]["path"]
+            cookbook_lock["source_options"].tap do |opt|
+              if cookbook_lock.has_key?("scm_info")
+                # Mutating path to rel is probably not the right thing to do.
+                opt["rel"] = opt["path"] unless opt["path"] == "."
+                opt.delete("path")
+                opt["git"]      = cookbook_lock["scm_info"]["remote"]
```

Comment:
@jaym @danielsdeleo  Should this key be `opt["remote"]` instead of `opt["git"]`?